cmd/dagger/functions.go

```diff
@@ -0,0 +1,546 @@
+// callContext holds the execution context for a function call.
+type callContext struct {
+	q      *querybuilder.Selection
+	e      *client.Client
+	c      graphql.Client
+	mod    *moduleDef
+	loader *progrock.VertexRecorder
+}
+
+func (c *callContext) Select(name string) {
+	c.q = c.q.Select(gqlFieldName(name))
+}
+
+func (c *callContext) Arg(name string, value any) {
+	c.q = c.q.Arg(gqlArgName(name), value)
+}
+
+func setCmdOutput(cmd *cobra.Command, vtx *progrock.VertexRecorder) {
+	if silent {
+		return
+	}
+
+	var stdout io.Writer
+	var stderr io.Writer
+
+	if stdoutIsTTY {
+		stdout = vtx.Stdout()
+	} else {
+		stdout = os.Stdout
+	}
+
+	if stderrIsTTY {
+		stderr = vtx.Stderr()
+	} else {
+		stderr = os.Stderr
+	}
+
+	cmd.SetOut(stdout)
+	cmd.SetErr(stderr)
+}
+
+// FuncCommand is a config object used to create a dynamic set of commands
+// for querying a module's functions.
+type FuncCommand struct {
+	// The name of the command (or verb), as shown in usage.
+	Name string
+
+	// Short is the short description shown in the 'help' output.
+	Short string
+
+	// Example is examples of how to use the command.
+	Example string
+
+	// Execute circumvents the default behavior of adding subcommands and
+	// executing them.
+	Execute func(*callContext, *cobra.Command) error
+
+	// OnInit is called when the command is created and initialized, before execution.
+	OnInit func(*cobra.Command)
+
+	// OnSetFlags is called when the flags have finished being set for a subcommand.
+	OnSetFlags func(*cobra.Command, *modTypeDef) error
+
+	// OnSelectObject is called when building the query on a function that
+	// returns a "leaf" object (e.g., Container). This is useful to add an
+	// additional field selection to the object. If the new selection changes
+	// the return type of the function, return the new type definition.
+	OnSelectObject func(*callContext, string) (*modTypeDef, error)
+
+	// OnSelectObjectList is called when building the query on a function that
+	// returns a list of objects. This is useful to add an additional field
+	// selection to the objects. If the new selection changes the return type
+	// of the function, return the new type definition.
+	OnSelectObjectList func(*callContext, *modObject) (*modTypeDef, error)
+
+	// OnSelectFunc is called after making a selection on a function to build
+	// the query. This can be useful to validate the return type or make
+	// any last minute changes to the query. If the new selection changes the
+	// return type of the function, return the new type definition.
+	OnSelectFunc func(*callContext, *modTypeDef) (*modTypeDef, error)
+
+	// CheckReturnType is called before the query is built, to validate
+	// the return type of the function. Can also be used as a last effort
+	// to select a sub-field.
+	CheckReturnType func(*callContext, *modTypeDef) error
+
+	// OnResult is called when the query has completed and returned a result.
+	OnResult func(*callContext, *cobra.Command, modTypeDef, *any) error
```

Comment:
nit: is `*any` needed vs. just `any`?

Reply:
Oh yes, def. a mistake, passed by my attention multiple times 🙂